### PR TITLE
Catch exceptions when retrieving data from Activity

### PR DESCRIPTION
**Pull request by yangyansong-adbe**
Port a bug fix (https://github.com/adobe/aepsdk-core-android/pull/658) from 2.x to 3.x

### CODE REVIEW

**Comment by praveek** [comment]
Can we add some tests by mocking Bundle?

**Comment by yangyansong-adbe** [comment]
> Can we add some tests by mocking Bundle?

I tried to mock the Bundle object with Mockito, but it doesn't work for some reason. Below is the sample code: 
```kotlin
val intent =
    Intent(ApplicationProvider.getApplicationContext(), TestActivity::class.java).apply {
        putExtra("exceptionKey", "exception")
        putExtra("key", "value")
    }

val spiedBundle = Mockito.spy(Bundle())
Mockito.`when`(spiedBundle.get("exceptionKey")).thenThrow(RuntimeException("Test Exception"))
val spiedIntent = Mockito.spy(intent)
Mockito.`when`(spiedIntent.extras).thenReturn(spiedBundle)
val activity = activityTestRule.launchActivity(spiedIntent)
val result = DataMarshaller.marshal(activity)
assertEquals(
    mapOf("key" to "value"),
    result
)
```
@praveek  Let me know if you have other suggestions.

**Comment by prudrabhat** [comment]
> > Can we add some tests by mocking Bundle?
> 
> I tried to mock the Bundle object with Mockito, but it doesn't work for some reason. Below is the sample code:
> 
> ```kotlin
> val intent =
>     Intent(ApplicationProvider.getApplicationContext(), TestActivity::class.java).apply {
>         putExtra("exceptionKey", "exception")
>         putExtra("key", "value")
>     }
> 
> val spiedBundle = Mockito.spy(Bundle())
> Mockito.`when`(spiedBundle.get("exceptionKey")).thenThrow(RuntimeException("Test Exception"))
> val spiedIntent = Mockito.spy(intent)
> Mockito.`when`(spiedIntent.extras).thenReturn(spiedBundle)
> val activity = activityTestRule.launchActivity(spiedIntent)
> val result = DataMarshaller.marshal(activity)
> assertEquals(
>     mapOf("key" to "value"),
>     result
> )
> ```
> 
> @praveek Let me know if you have other suggestions.

@yangyansong-adbe - Since Bundle is an Android component you will have to use `RobolectricTestRunner` or write it as an instrumentation test. Did you try any of these approaches?

**Comment by yangyansong-adbe** [comment]
> > > Can we add some tests by mocking Bundle?
> > 
> > 
> > I tried to mock the Bundle object with Mockito, but it doesn't work for some reason. Below is the sample code:
> > ```kotlin
> > val intent =
> >     Intent(ApplicationProvider.getApplicationContext(), TestActivity::class.java).apply {
> >         putExtra("exceptionKey", "exception")
> >         putExtra("key", "value")
> >     }
> > 
> > val spiedBundle = Mockito.spy(Bundle())
> > Mockito.`when`(spiedBundle.get("exceptionKey")).thenThrow(RuntimeException("Test Exception"))
> > val spiedIntent = Mockito.spy(intent)
> > Mockito.`when`(spiedIntent.extras).thenReturn(spiedBundle)
> > val activity = activityTestRule.launchActivity(spiedIntent)
> > val result = DataMarshaller.marshal(activity)
> > assertEquals(
> >     mapOf("key" to "value"),
> >     result
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @praveek Let me know if you have other suggestions.
> 
> @yangyansong-adbe - Since Bundle is an Android component you will have to use `RobolectricTestRunner` or write it as an instrumentation test. Did you try any of these approaches?
@prudrabhat  
It's an instrumentation test. Now, we can use mockito to mock Android components in instrumentation tests after adding `com.linkedin.dexmaker:dexmaker-mockito-inline`. 
I haven't tried Robolectric because it would require rewriting the entire `DataMarshallerTests` class.

**Comment by navratan-soni** [comment]
> > > > Can we add some tests by mocking Bundle?
> > > 
> > > 
> > > I tried to mock the Bundle object with Mockito, but it doesn't work for some reason. Below is the sample code:
> > > ```kotlin
> > > val intent =
> > >     Intent(ApplicationProvider.getApplicationContext(), TestActivity::class.java).apply {
> > >         putExtra("exceptionKey", "exception")
> > >         putExtra("key", "value")
> > >     }
> > > 
> > > val spiedBundle = Mockito.spy(Bundle())
> > > Mockito.`when`(spiedBundle.get("exceptionKey")).thenThrow(RuntimeException("Test Exception"))
> > > val spiedIntent = Mockito.spy(intent)
> > > Mockito.`when`(spiedIntent.extras).thenReturn(spiedBundle)
> > > val activity = activityTestRule.launchActivity(spiedIntent)
> > > val result = DataMarshaller.marshal(activity)
> > > assertEquals(
> > >     mapOf("key" to "value"),
> > >     result
> > > )
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > @praveek Let me know if you have other suggestions.
> > 
> > 
> > @yangyansong-adbe - Since Bundle is an Android component you will have to use `RobolectricTestRunner` or write it as an instrumentation test. Did you try any of these approaches?
> 
> @prudrabhat It's an instrumentation test. Now, we can use mockito to mock Android components in instrumentation tests after adding `com.linkedin.dexmaker:dexmaker-mockito-inline`. I haven't tried Robolectric because it would require rewriting the entire `DataMarshallerTests` class.

Can we user something like this, I have tried mocking the Intent and bundle as well, and it is working for me in one of the test cases, not sure if this fits the requirements.
`val mockIntent = mock<Intent>()
        val mockBundle = mock<Bundle>()
        Mockito.`when`(mockBundle.getString(PushTemplateConstants.IntentKeys.TITLE_TEXT))
            .thenReturn(MOCKED_TITLE)
        Mockito.`when`(mockBundle.getString(PushTemplateConstants.IntentKeys.BODY_TEXT))
            .thenReturn(MOCKED_BODY)
        Mockito.`when`(mockBundle.getInt(PushTemplateConstants.IntentKeys.PAYLOAD_VERSION))
            .thenReturn(MOCKED_PAYLOAD_VERSION)
        `when`(mockIntent.extras).thenReturn(mockBundle)`

File: code/core/src/main/java/com/adobe/marketing/mobile/internal/DataMarshaller.kt [hunk 1]
```diff
@@ -70,9 +70,13 @@ internal object DataMarshaller {
...
-            val value = extraBundle[key]
-            if (value?.toString()?.isNotEmpty() == true) {
-                marshalledData[newKey] = value
+            try {
+                val value = extraBundle[key]
+                if (value?.toString()?.isNotEmpty() == true) {
+                    marshalledData[newKey] = value
+                }
+            } catch (e: Exception) {
+                Log.error(CoreConstants.LOG_TAG, LOG_TAG, "Failed to retrieve data (key = $key) from Activity, error is: ${e.message}")
```
Comment (prudrabhat) [inline code/core/src/main/java/com/adobe/marketing/mobile/internal/DataMarshaller.kt:79]: I think we can to test this flow without mocks as follows : 

```
    class ObjectThrowsOnToString : Serializable {
        override fun toString(): String {
            throw IllegalStateException("This is a test exception")
        }
    }

    @Test
    fun testMarshalIntentExtras_CatchesExceptionsDuring() {
        val throwsException = ObjectWithToString()

        val intent = Intent(ApplicationProvider.getApplicationContext(), TestActivity::class.java).apply {
            // Add an extra to the bundle with a key that will cause an exception when trying to retrieve it
            putExtra("testKey", throwsException)
        }

        val activity = activityTestRule.launchActivity(intent)

        try {
            val returned = DataMarshaller.marshal(activity)
            // check returned map here
        } catch (e: Exception) {
            fail("Exception should have been caught")
        }
    }

```

Comment (yangyansong-adbe) [reply code/core/src/main/java/com/adobe/marketing/mobile/internal/DataMarshaller.kt:79]: @prudrabhat @navratan-soni thank you for the suggestions. I tried the "without mocks" method, and it works great.

**Pull request merged**